diff --git a/src/tests.c b/src/tests.c
index c2d5e28..7fc5b4e 100644
--- a/src/tests.c
+++ b/src/tests.c
@@ -5379,6 +5379,10 @@ void run_ecdsa_openssl(void) {
 }
 #endif
 
+#ifdef ENABLE_MODULE_AESNI
+# include "modules/aesni/tests_impl.h"
+#endif
+
 #ifdef ENABLE_MODULE_ECDH
 # include "modules/ecdh/tests_impl.h"
 #endif
@@ -5653,6 +5657,11 @@ int main(int argc, char **argv) {
     /* EC key arithmetic test */
     run_eckey_negate_test();
 
+#ifdef ENABLE_MODULE_AESNI
+    /* aesni tests */
+    run_aesni_tests();
+#endif
+
 #ifdef ENABLE_MODULE_ECDH
     /* ecdh tests */
     run_ecdh_tests();
