diff --git a/src/valgrind_ctime_test.c b/src/valgrind_ctime_test.c
index cfca5a1..378a67c 100644
--- a/src/valgrind_ctime_test.c
+++ b/src/valgrind_ctime_test.c
@@ -11,6 +11,10 @@
 #include "assumptions.h"
 #include "util.h"
 
+#ifdef ENABLE_MODULE_AESNI
+# include "include/secp256k1_aesni.h"
+#endif
+
 #ifdef ENABLE_MODULE_ECDH
 # include "include/secp256k1_ecdh.h"
 #endif
@@ -100,6 +104,11 @@ void run_tests(secp256k1_context *ctx, unsigned char *key) {
     CHECK(ret);
     CHECK(secp256k1_ecdsa_signature_serialize_der(ctx, sig, &siglen, &signature));
 
+#ifdef ENABLE_MODULE_AESNI
+    /* Test AESNI. */
+    CHECK(1 == 1);
+#endif
+
 #ifdef ENABLE_MODULE_ECDH
     /* Test ECDH. */
     VALGRIND_MAKE_MEM_UNDEFINED(key, 32);
